language/sygus.tex

```diff
@@ -2632,6 +2730,90 @@ \section{Examples}%
 and hence the solution for $f$ satisfies all input constraints.
 \end{example}
 
+
+\begin{example}[Optimization with Weights]
+Consider the following example:
+\begin{lstlisting}[language=SyGuS]
+(set-logic LIA)
+(set-feature :weights true)
+(declare-weight branches)
+(synth-fun f ((x Int)) Int
+   ((I Int) (B Bool))
+   ((I Int (0 1 x (+ I I) (! (ite B I I) :branches 1)))
+    (B Bool ((>= I I) (= I I)))))
+(declare-var x Int)
+(constraint (= (f x) (ite (= x 0) 0 x)))
+(optimize-synth ((! (_ branches f) :min)))
+\end{lstlisting}
+In this example, a weight keyword $\texttt{branches}$ has been declared.
+The grammar of function $\texttt{f}$ has been annotated such that
+$\texttt{branches}$ corresponds to the number of $\texttt{ite}$ terms in its body.
+An optimization query is made, where
+the objective is to minimize the number of branches in $\texttt{f}$.
+This is specified by a term annotation on $\texttt{(\_ branches f)}$.
+A possible correct response for this example is:
+\begin{lstlisting}[language=SyGuS]
+(
+(1)
+(define-fun f ((x Int)) Int (ite (= x 0) 0 x))
+)
+\end{lstlisting}
+Another possible response for this example is:
+\begin{lstlisting}[language=SyGuS]
+(
+(0)
+(define-fun f ((x Int)) Int x)
+)
+\end{lstlisting}
+In both cases, the value of term $\texttt{(\_ branches f)}$ was given
+in the solution that was consistent with the number of
+$\texttt{ite}$ terms in the body of $\texttt{f}$.
+Hence, both responses are correct.
+The latter solution gives a smaller value for this term,
+and hence it is preferred based on the specified objective.
+
+Notice that in this example,
+no attribute is provided for the $\texttt{optimize-synth}$ command,
+meaning that no ordering is specified on the tuple passes as argument
+to this command.
+For details on the default semantics for objectives, see \cref{sec:objectives}.
+In the next example, we show how ordering on tuples can be specified.
+\end{example}
+
+\begin{example}[Lexicographic Optimization]
+Consider the following example:
+\begin{lstlisting}[language=SyGuS]
+(set-logic LIA)
+(synth-fun f ((x Int)) Int
+   ((I Int) (B Bool))
+   ((I Int (0 1 x (+ I I) (ite B I I)))))
+(declare-var x Int)
+(constraint (or (= (f x) 1) (= (f x) x)))
+(optimize-synth ((! (f 0) :max) (! (f 100) :max)) :lexicographic)
```

Comment:
In both examples (this one and above) we are using attribute-annotated terms. Would it make sense to use `optimize-synth` on base terms that are not arribute-annotated?

For instance, would

```smt
(optimize-synth (f 0) (f 100) :lexicographic)
```

mean anything?

Reply:
May be each term would have a singleton value and there wouldn't be anything to optimize? Just thinking out loud ...

Reply:
first, notice that the syntax always expects a tuple, so you'd have to do:
```
(optimize-synth ((f 0) (f 100)) :lexicographic)
```
For unannotated terms, we would have to settle on a default ordering, which I'll specify in the objectives section.

Reply:
I understand now, thanks